SFML/SFML
# Fix order and remove duplicates of KeyPressed events

## Description

This is an attempt to fix broken keyboard events on X11 with IBus input method.

What this PR fixes (hopefully):
1. Avoid duplicated `KeyPressed` events.
   This is an alternative to #1894 from which I took inspiration, but more robust by indexing the bitset with X11 keycodes instead of `sf::Keyboard::Key` enumeration.

2. Fix events being in the wrong order, i.e. fixes #2191
   I noticed that getting as many events as possible from X11 queue and only then processing them (behavior added in #1291) makes the input method filtering go wrong, especially when framerate is low.
   Popping and processing events in order, with some contortions to handle key repeat events, seems to work just great.
   There is a risk of reintroducing bug #1223, but I do not reproduce that bug when testing this present PR.

What this PR does not fix, but improves a little bit:
- `TextEntered` events coming slowly when framerate is low.
  As far as I understand, when IBus input method is used and a key is pressed, SFML pops a `KeyPress` event from X11 events queue and notifies the input method about it through `XFilterEvent`. The input methods gives an answer with text after a tiny delay through another X11 `KeyPress` event, so it is available for SFML on the next frame. This delay is noticeable when framerate is low.
   Before this PR, when holding a key down with a low framerate, I had one `TextEntered` event every two frames. Somehow with this PR, I get one `TextEntered` each frame. So the problem becomes less noticeable.
   A workaround to avoid this delay could be to process events in a separate thread but that would add complexity and I am not sure X11 would not explode.

## Tasks

* [ ] Get more people to test @Edgaru089 @Neko-Box-Coder

## How to test this PR?

### Test program

```c++
#include <iostream>
#include <SFML/Graphics.hpp>

int main(int argc, char* argv[])
{
    // usage: test <framerate> <keyRepeat>
    const int framerate = 1 < argc ? std::atoi(argv[1]) : 60;
    const bool keyRepeat = 2 < argc ? std::atoi(argv[2]) : true;

    sf::Window window(sf::VideoMode(100, 100), "Window");

    auto createWindow = [&] {
        window.create(sf::VideoMode(100, 100), "Window");
        window.setFramerateLimit(framerate);
        window.setKeyRepeatEnabled(keyRepeat);
    };
    createWindow();

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
            else if (event.type == sf::Event::KeyPressed)
            {
                std::cout << "Key Pressed: " << event.key.code << std::endl;
                if (event.key.code == sf::Keyboard::Key::Space)
                    createWindow();
            }
            else if (event.type == sf::Event::KeyReleased)
                std::cout << "Key Released: " << event.key.code << std::endl;
            else if (event.type == sf::Event::TextEntered)
            {
                const auto utf8 = sf::String(event.text.unicode).toUtf8();
                std::cout << "Text Entered: " << std::string(utf8.begin(), utf8.end()) << std::endl;
            }
        }

        window.display();
    }
}
```

### Building the test program

Save the `main.cpp` file above in a directory along with the following `CMakeLists.txt` file:
```cmake
cmake_minimum_required(VERSION 3.16)
project(test)

include(FetchContent)
FetchContent_Declare(SFML
    GIT_REPOSITORY https://github.com/kimci86/SFML.git
    GIT_TAG bugfix/x11_key_events)
if(NOT sfml_POPULATED)
  FetchContent_Populate(SFML)
  add_subdirectory(${sfml_SOURCE_DIR} ${sfml_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

add_executable(test main.cpp)
target_link_libraries(test PRIVATE sfml-window)
target_compile_features(test PRIVATE cxx_std_17)
```

Then run those commands from the directory the files are saved in:
```sh
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build
```

### Running the test program

From the same directory:
```sh
# by default, running at 60 fps with key repeat enabled
build/test
# running at 20 fps with key repeat enabled
build/test 20
# running at 20 fps with key repeat disabled
build/test 20 0
```

### Testing steps

Using the test program, with different framerates (first command line argument):
- Check that a single `KeyPressed` event is generated when pressing a key for a short time.
- Check that a single `KeyReleased` event is generated and comes after all the `KeyPressed` events when pressing a key for a long time.
- With key repetition disabled (second command line argument set to 0), check that a long press generates the same events as a short press.
  + Check that it behaves the same after recreating the window by pressing the space bar.
